processor/src/stack/overflow.rs

```diff
@@ -1,211 +1,282 @@
+        self.clk += 1;
     }
+}
+
+/// Stores the history of the overflow stack at every clock cycle, where only the relevant context
+/// is stored in the history for each clock cycle.
+///
+/// The events which update the overflow stack history are:
+/// - `push` operation,
+/// - `pop` operation,
+/// - a new context is started,
+/// - a former context is restored.
+#[derive(Debug)]
+struct OverflowStackHistory {
+    /// Stores the full state of the overflow stack at every clock cycle. Formally, if `clk <
+    /// last_transition_clk`, then there is exactly one range in the history which contains `clk`.
+    history: Vec<(RangeInclusive<RowIndex>, Vec<Felt>)>,
+}
 
-    /// Returns the number of overflowing stack elements at the current clock cycle.
-    pub fn num_active_rows(&self) -> usize {
-        self.active_rows.len()
+impl OverflowStackHistory {
+    /// Creates a new empty overflow stack history.
+    pub fn new() -> Self {
+        Self { history: Vec::new() }
     }
 
-    // HELPER METHODS
+    // PUBLIC ACCESSORS
     // --------------------------------------------------------------------------------------------
 
-    /// Saves a copy of the current table state into the trace at the specified clock cycle.
-    fn save_current_state(&mut self, clk: u64) {
-        debug_assert!(self.trace_enabled, "overflow table trace not enabled");
-        let current_state = self.active_rows.iter().map(|&idx| self.all_rows[idx].val).collect();
-        self.trace.insert(clk, current_state);
+    /// Returns the overflow stack at the given clock cycle in the active context at that clock
+    /// cycle if present in the history, otherwise returns `None`.
+    ///
+    /// That is, if the target clock cycle is greater than the last clock cycle at which the
+    /// history was updated, the method returns `None`.
+    ///
+    /// The first element returned by the iterator is the top of the overflow stack.
+    pub fn get_at(&self, target_clk: RowIndex) -> Option<impl Iterator<Item = &Felt>> {
+        match self.last_clk_in_history() {
+            Some(last_clk_in_history) => {
+                if target_clk > last_clk_in_history {
+                    None
+                } else {
+                    for (range, state) in self.history.iter() {
+                        if range.contains(&target_clk) {
+                            return Some(state.iter().rev());
+                        }
+                    }
+
+                    unreachable!("overflow stack history not properly constructed")
+                }
+            },
+            None => None,
+        }
     }
 
-    // TEST ACCESSORS
+    // PUBLIC MUTATORS
     // --------------------------------------------------------------------------------------------
 
-    #[cfg(test)]
-    pub fn all_rows(&self) -> &[OverflowTableRow] {
-        &self.all_rows
+    /// Indicates that an element was added to or removed from the overflow stack at the current
```

Comment:
nit: Would "Registers" or "Logs" be better than "Indicates" to be better convey the meaning?

Reply:
Switched to "Registers"